### PR TITLE
[system-probe][NET-381] scope dns stats by domain

### PR DESCRIPTION
**Pull request by mbakht**
### What does this PR do?

This adds the option of scoping DNS stats by domain.

The `dnsStats` map earlier was one layer : `dnsKey` -> `dnsStats`
This PR introduces an additional layer : `dnsKey` -> `domain` -> `dnsStats`

So far, no visible difference has been noticed in CPU usage (the yellow one is the with domains):

Screen Shot 2020-11-06 at 9 36 48 AM

### Gating

The feature is gated by `system_probe_config.collect_dns_domains` and the environment variable `DD_COLLECT_DNS_DOMAINS`.

Initially, I was planning to keep the codepath and data structures for **domain-scoped** stats completely separate from the existing **domain-agnostic** one.

However, that led to too much complexity
-  We could have two versions of `DNSStatKeeper` and use one based on config variable.
-  But then we would have to have duplicate data structures in `State` to support both types of data.

So this PR takes a different approach. It makes all the data structures and functions  **domain-scoped**, even when the config flag is turned off.  This is what happens differently based on whether the flag is turned off or not:
-  `dnsParser` leaves the domain field `nil`, so **all** stats get scoped by the same domain
```go
if p.collectDNSDomains {
    pktInfo.question = string(question.Name)
}
```
- `networkState`   fills up  (`conn.DNSSuccessfulResponses`, `conn.DNSTimeouts`) OR `conn.DNSStatsByDomain[domain]`

```go
for domain, dnsStats := range dnsStatsByDomain {
    if ns.collectDNSDomains {
        var ds DNSStats
	ds.DNSTimeouts = dnsStats.timeouts
        ...... 
	conn.DNSStatsByDomain[domain] = ds
    } else {
	conn.DNSSuccessfulResponses += dnsStats.countByRcode[DNSResponseCodeNoError]
	conn.DNSTimeouts += dnsStats.timeouts
        ......
   }
```

### Encoding

It is likely that the set of unique domains queried would be really small. But if store them separately for each connection, it would be lots of duplication.
Initially, we looked at reusing `encodedDNS` field which stores the `IP->domain` mapping.
But that approach was abandoned because of the complexity:
- The set of domains would not be same. For example, if a query for a domain resulted in `NXDomain`, it will show up in `DNSStats` but not in the `IP->domain` mapping. So injecting those extra domains would make the code complex.
- The pointers would have to be updated in each connection after the generation of  `encodedDND`

Instead, we simply have a new field (a list of `string`s) that contains all the domains and `DNSStats` stores pointers to positions in that list:
```
    "dnsStatsByDomain": {
        "0": {
          "dnsTimeouts": 0,
          "dnsSuccessLatencySum": "18505",
          "dnsFailureLatencySum": "0",
          "dnsCountByRcode": {
            "0": 1
          }
        }
      }
    "dnsStatsByDomain": {
        "1": {
          "dnsTimeouts": 0,
          "dnsSuccessLatencySum": "20261",
          "dnsFailureLatencySum": "0",
          "dnsCountByRcode": {
            "0": 1
          }
        }
  "domains": [
    "cnn.com",
    "bbc.com"
  ]
```

#### Next steps for encoding

To reuse `encodedDNS` we plan to take the following approach:

1. Go through all `connections` and for each `connection`, go through all `DNSStats` and create a set of `domains` (`pkg/network/encoding/format.go`, `pkg/network/encoding/protobuf.go`, `pkg/network/encoding/json.go`)

2. Change `dnsEncoder.encode` (https://github.com/DataDog/datadog-agent/blob/master/pkg/process/checks/net.go#L209) and pass that list of strings

3. From `encodedDNS`,  create a map  from domain->position.

4. Go through all `connections` and for each `connection`, go through all `DNSStats` and update the value for domain with the position in the encoded value.

### Describe your test plan

1. Make sure collections of `domains` is enabled:
```
system_probe_config:
  enabled: true
  collect_dns_domains: true
```
2. Run `system-probe`. `Dig` a few domains.
- `dig apple.com`
- `dig datadoghq.com`

3. Run the following command `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | python parser.py` where `parser.py` has the following contents: 

```python
import json
import sys

d = json.load(sys.stdin)
print "----------------------------------------------------------------------------"
for c in d['conns']:
    if not c['dnsStatsByDomain']:
        continue
    s =  "{: <5} {} {: >12}:{: <5} ->{: >16}:{: <5}"
    print s.format(
        c['pid'],
        c['type'],
        c['laddr']['ip'],
        c['laddr']['port'],
        c['raddr']['ip'],
        c['raddr']['port'],
    )
    for k, v in c['dnsStatsByDomain'].iteritems():
        s =  "{: >20} {: >2} {: >2} {: >2} {: >2} {: >2}"
        ec = v['dnsCountByRcode']
        success = 0
        failure = 0
        for codestr, count in ec.iteritems():
            code = int(codestr)
            if code == 0:
                success += 1
            else:
                failure += 1

        print s.format(
           d['domains'][int(k)],
           success,
           failure,
           v['dnsTimeouts'],
           v['dnsSuccessLatencySum'],
           v['dnsFailureLatencySum'],
        )
    print "----------------------------------------------------------------------------"
```
The output should be something like the following:

```
----------------------------------------------------------------------------
8199  udp    10.0.2.15:55238 ->        10.0.2.3:53
           apple.com  1  0  0 34892  0
----------------------------------------------------------------------------
8199  udp    10.0.2.15:36959 ->        10.0.2.3:53
       datadoghq.com  1  0  0 39834  0
----------------------------------------------------------------------------
```

### CODE REVIEW

File: pkg/process/config/yaml_config.go [hunk 1]
```diff
@@ -50,6 +50,12 @@ func (a *AgentConfig) loadSysProbeYamlConfig(path string) error {
 		a.CollectDNSStats = config.Datadog.GetBool(key(spNS, "collect_dns_stats"))
 	}
 
+	if config.Datadog.IsSet(key(spNS, "collect_dns_domains")) {
+		fmt.Println("Setting collect dns domains")
```
Comment (hmahmood) [inline pkg/process/config/yaml_config.go:54]: Are the `Println`'s left over from debugging?

File: pkg/network/dns_stats.go [hunk 2]
```diff
@@ -157,12 +163,20 @@ func (d *dnsStatKeeper) removeExpiredStates(earliestTs time.Time) {
...
-			d.deleteCount++
-			stats := d.getStats(k.key)
+			allStats, ok := d.stats[k.key]
+			if !ok {
+				allStats = make(map[string]dnsStats)
+			}
+			stats, ok := allStats[v.question]
+			if !ok {
+				stats.countByRcode = make(map[uint8]uint32)
+				d.deleteCount++
```
Comment (hmahmood) [inline pkg/network/dns_stats.go:175]: Previously this was incremented regardless of whether stats existed for `k.key` or not, why the change?

File: pkg/network/encoding/format.go [hunk 3]
```diff
@@ -21,7 +21,22 @@ var connPool = sync.Pool{
...
+		ms.DnsSuccessLatencySum = s.DNSSuccessLatencySum
+		ms.DnsTimeouts = s.DNSTimeouts
+		pos, ok := domainSet[d]
+		if !ok {
+			pos = len(domainSet)
+			domainSet[d] = pos
+		}
+		m[int32(pos)] = &ms
+	}
+
```
Comment (hmahmood) [inline pkg/network/encoding/format.go:39]: Could you move this to something like `formatDNSStatsByDomain()` to maintain the pattern for the other attributes in this function?

File: pkg/network/encoding/format.go [hunk 4]
```diff
@@ -47,6 +62,7 @@ func FormatConnection(conn network.ConnectionStats) *model.Connection {
 	c.DnsCountByRcode = conn.DNSCountByRcode
 	c.LastTcpEstablished = conn.LastTCPEstablished
 	c.LastTcpClosed = conn.LastTCPClosed
+	c.DnsStatsByDomain = m
```
Comment (leeavital) [inline pkg/network/encoding/format.go:65]: I don't think we should  include both DnsFailureLatencySum and DnsStatsByDomain in they have the same information. Can we include only one based on whether or not domain collection is enabled  (with the understanding that we will need some extra logic in the backend to handle agents where this field is set)

Comment (leeavital) [reply pkg/network/encoding/format.go:65]: never mind, I missed that this was being done in state.go

File: pkg/network/encoding/json.go [hunk 5]
```diff
@@ -17,12 +17,20 @@ type jsonSerializer struct {
...
+	}
+
+	domains := make([]string, len(domainSet))
+	for k, v := range domainSet {
+		domains[v] = k
 	}
 
 	payload := connsPool.Get().(*model.Connections)
 	payload.Conns = agentConns
+	payload.Domains = domains
```
Comment (leeavital) [inline pkg/network/encoding/json.go:33]: Will we be encoding the same domains in FormatDNS and `payload.Domains` ?

File: pkg/process/config/yaml_config.go (hunk 1 above)
Comment (mbakht) [reply pkg/process/config/yaml_config.go:54]: Yes 🤦 . Will remove. Thanks!

File: pkg/network/dns_stats.go (hunk 2 above)
Comment (mbakht) [reply pkg/network/dns_stats.go:175]: The change was unintentional. Thanks for catching!

File: pkg/network/encoding/json.go (hunk 5 above)
Comment (mbakht) [reply pkg/network/encoding/json.go:33]: Yes, there will be some overlap but the sets won't always be the same. For example, domains that don't receive successful replies won't be be part of `payload.Dns` but they will be part of `payload.Domains`.

**Review by hmahmood — APPROVED** [review]

**Comment by mbakht** [comment]
Based on my discussion with @leeavital, we have decided not to merge this for `7.25` and explore options for merging different DNS data.

File: pkg/process/checks/net_test.go [hunk 6]
```diff
@@ -175,3 +175,44 @@ func TestBatchSimilarConnectionsTogether(t *testing.T) {
 	}
 	assert.Equal(t, 6, total)
 }
+
+func TestNetworkConnectionBatchingWithDomains(t *testing.T) {
+	p := makeConnections(4)
```
Comment (leeavital) [inline pkg/process/checks/net_test.go:180]: nit: nit let's call this `conns`

**Review by leeavital — APPROVED** [review]

**Review by derekwbrown — APPROVED** [review]

**Comment by derekwbrown** [comment]
Merging absent `agent-core` approval, as was required only for `go.mod`
